### PR TITLE
Improve numerical stability of `torch.distributions.wishart.Wishart`

### PR DESCRIPTION
**Pull request by nonconvexopt**
Maintanance of https://github.com/pytorch/pytorch/pull/70377
Multiple modifications of the merged initial implementation of Wishart distribution.

Key modifications:
* torch/distributions/wishart.py: Clamp (Clip) float type values to calculate reciprocal in numerically stable manner, by using the eps value paired to each torch.dtype
* test/distributions/test_distributions.py: Test Wishart distribution implementation in numerically unstable zones, i.e df values are at ndim - 1 < df < ndim where ndim is the one dimenstion of the Wishart parameter & sample matrix.

Re-opened reverted PR #72059
cc @neerajprad @vadimkantorov

### CODE REVIEW

**Comment by nonconvexopt** [comment]
ciflow/all

**Comment by nonconvexopt** [comment]
label: ciflow/all

File: torch/distributions/wishart.py [hunk 1]
```diff
@@ -20,15 +20,20 @@ def _mvdigamma(x: torch.Tensor, p: int) -> torch.Tensor:
         - torch.arange(p, dtype=x.dtype, device=x.device).div(2).expand(x.shape + (-1,))
     ).sum(-1)
 
+def _clamp_with_eps(x: torch.Tensor) -> torch.Tensor:
```
Comment (vadimkantorov) [inline torch/distributions/wishart.py:23]: related: https://github.com/pytorch/pytorch/issues/41492, this might not be torchscriptable...

Comment (nonconvexopt) [reply torch/distributions/wishart.py:23]: Thank you for letting me know. I will try to find alternatives. It is hard to judge which function is torchscriptable.

Comment (vadimkantorov) [reply torch/distributions/wishart.py:23]: Also feel free to chime in that issue, maybe its priority might be increased

Comment (nonconvexopt) [reply torch/distributions/wishart.py:23]: I don't think I understand the mechanism of `torch.script` yet, but I will try on. Thanks for the suggestion!

**Review by fritzo — COMMENTED** [review]
Hi @nonconvexopt, while we appreciate cleanup PRs, this PR mixes cleanup with functional changes, making it challenging for me to distinguish mere cleanup from the core relevant fixes in this PR.  In the future consider separating PRs to accomplish either cleanup or behavior change.  When PRs are split into small coherent parts, they are less susceptible to reversion and are easier to review.

In this PR could you clarify some questions of mine?

1. It looks like the `._natural_params()` are being changed and reordered. Can you help me understand the change here?  Why are they being reordered?  Why is there a shift of `(p+1)/2` which is then removed in `._log_normalizer()`?  Wouldn't it be cheaper to avoid this shift in the first place?
2. Can you confirm that `torch.mvlgamma(x,p) = _mvdigamma(x, p)` in your `_log_normalizer()` changes?  If so, why not simply remove `_mvdigamma()` from this file?  Does this have anything to do with numerical stability, or is it simply cleanup, or is it a bug fix?

File: torch/distributions/wishart.py (hunk 1 above)
Comment (nonconvexopt) [reply torch/distributions/wishart.py:23]: Thanks for suggestion. I will modify the name.

**Comment by nonconvexopt** [comment]
In the [Wikepedia Exponential Family page](https://en.m.wikipedia.org/wiki/Exponential_family), natural parameter of Wishart distribution is noted as: `-1/2 * V^{-1}` and `(n-p-1)/2`

**Comment by nonconvexopt** [comment]
@neerajprad @vadimkantorov @fritzo I think I am not allowed to reproduce the macos errors. Does the waiting for the further maintenance of CI will solve the issue?

```
botocore.exceptions.ClientError: An error occurred (AccessDeniedException) when calling the Invoke operation: User: arn:aws:iam::308535385114:user/pytorch-pytorch-ossci-s3-read-write is not authorized to perform: lambda:InvokeFunction on resource: arn:aws:lambda:us-east-1:308535385114:function:rds-proxy because no identity-based policy allows the lambda:InvokeFunction action
Error: Process completed with exit code 1.
```

File: torch/distributions/wishart.py [hunk 2]
```diff
@@ -13,12 +13,9 @@
...
 
-def _mvdigamma(x: torch.Tensor, p: int) -> torch.Tensor:
-    assert x.gt((p - 1) / 2).all(), "Wrong domain for multivariate digamma function."
-    return torch.digamma(
-        x.unsqueeze(-1)
-        - torch.arange(p, dtype=x.dtype, device=x.device).div(2).expand(x.shape + (-1,))
-    ).sum(-1)
+def _clamp_above_eps(x: torch.Tensor) -> torch.Tensor:
+    # We assume positive input for this function
+    return x.clamp(min=torch.finfo(x.dtype).eps)
```
Comment (fritzo) [inline torch/distributions/wishart.py:18]: I'm not sure if the `test_bad_symbolic_registration` test failures are related, but I would think it should be safe to use `torch.finfo` in this helper.

File: torch/distributions/wishart.py [hunk 3]
```diff
@@ -250,32 +251,44 @@ def log_prob(self, value):
...
 
     @property
     def _natural_params(self):
+        nu = self.df  # has shape (batch_shape)
+        p = self._event_shape[-1]  # has singleton shape
+
         return (
-            0.5 * self.df,
             - 0.5 * self.precision_matrix,
+            0.5 * (nu - p - 1),
```
Comment (fritzo) [inline torch/distributions/wishart.py:286]: This looks like a correct but unnecessary change.

File: torch/distributions/wishart.py [hunk 4]
```diff
@@ -250,32 +251,44 @@ def log_prob(self, value):
...
     def entropy(self):
         nu = self.df  # has shape (batch_shape)
         p = self._event_shape[-1]  # has singleton shape
         V = self.covariance_matrix  # has shape (batch_shape x event_shape)
+
         return (
             (p + 1) * self._unbroadcasted_scale_tril.diagonal(dim1=-2, dim2=-1).log().sum(-1)
-            + p * (p + 1) * _log_2 / 2
-            + torch.mvlgamma(nu / 2, p=p)
-            - (nu - p - 1) / 2 * _mvdigamma(nu / 2, p=p)
```
Comment (fritzo) [inline torch/distributions/wishart.py:268]: Why did you inline the mvdigamma function? Wouldn't it be clearer to preserve it being defined as a separate function, so readers understood this complex expression is a multivariate digamma function or something?

File: torch/distributions/wishart.py [hunk 5]
```diff
@@ -250,32 +251,44 @@ def log_prob(self, value):
...
         )
 
     def _log_normalizer(self, x, y):
-        p = y.shape[-1]
-        return x * (- torch.linalg.slogdet(-2 * y).logabsdet + _log_2 * p) + _mvdigamma(x, p=p)
+        p = self._event_shape[-1]
+        return (
+            (y + 0.5 * (p + 1)) * (- torch.linalg.slogdet(-2 * x).logabsdet + _log_2 * p)
+            + torch.mvlgamma(y + 0.5 * (p + 1), p=p)
+        )
```
Comment (fritzo) [inline torch/distributions/wishart.py:294]: This looks like a correct but unnecessary change.

**Review by fritzo — APPROVED** [review]
LGTM. This is fine, but imho `x / 2` is easier to read than `0.5 * x` everywhere.  I believe test failures are unrelated (pip failures, onnx registration errors).

File: torch/distributions/wishart.py (hunk 3 above)
Comment (nonconvexopt) [reply torch/distributions/wishart.py:286]: You're right. I can clean this part.

File: torch/distributions/wishart.py (hunk 4 above)
Comment (nonconvexopt) [reply torch/distributions/wishart.py:268]: That would be great. I hope there is a official `torch.mvldigamma` function, but I am afraid that it will be only me. I will seperate the function as multivariate digamma function.

File: torch/distributions/wishart.py (hunk 5 above)
Comment (nonconvexopt) [reply torch/distributions/wishart.py:294]: I tried to make code neater and readable. I will try to make best outlooks.

File: torch/distributions/wishart.py (hunk 2 above)
Comment (nonconvexopt) [reply torch/distributions/wishart.py:18]: Thank you for letting me know. I am new to using torch.finfo. I understand that `torch.finfo` raise a issue in JIT for now.

**Comment by nonconvexopt** [comment]
I think it is ready to be merged.

**Review by fritzo — APPROVED** [review]
LGTM.

Test errors appear unrelated:

<details>

```
botocore.exceptions.ClientError: An error occurred (AccessDeniedException) when calling the Invoke operation: User: arn:aws:iam::308535385114:user/pytorch-pytorch-ossci-s3-read-write is not authorized to perform: lambda:InvokeFunction on resource: arn:aws:lambda:us-east-1:308535385114:function:rds-proxy because no identity-based policy allows the lambda:InvokeFunction action
```

</details>

@neerajprad would you be able to merge this?

**Comment by nonconvexopt** [comment]
Added trivial change: fix wrong parameter input (df with `ndim - 1 < df < ndim`) in Scipy Wishart class.

**Comment by nonconvexopt** [comment]
LGTM

**Review by fritzo — APPROVED** [review]
LGTM @neerajprad I believe this is ready to merge

**Comment by nonconvexopt** [comment]
Thank you for the richful feedbacks and reviews. I appreciated and kindness. I am considering implementing Multivariate T distribution for PyTorch if there is no one trying after I finish Inverse Wishart #70275. It's a shame that I am delaying it to this far.

I guess some people are trying to use T Process similary to Gaussian Process this days and some probabilistic ML libraries implemented it. It might be great if PyTorch offers official implementation for Multivariate T.

**Comment by fritzo** [comment]
@nonconvexopt it's a pleasure to work with you.  Yes Multivariate Student't t might be useful for T processes, but I'm not sure: IIUC it is just a Gamma-MultivariateNormal chain where the gamma rv can be analytically integrated out, and I believe in T processes there are local gammas, but in mutlivariate T there is a single gamma?  Let's move this discussion to a new issue.  Also cc @fehiepsi who worked out a lot of that math [in Pyro](https://github.com/pyro-ppl/pyro/pull/2233).

**Comment by fehiepsi** [comment]
There is [MVT](https://docs.pyro.ai/en/latest/distributions.html#multivariatestudentt) distribution in Pyro, but it only supports `scale_tril` input. If we port/reimplement in PyTorch, I think we need to support `covariance_matrix`, `precision_matrix` inputs too.
